src/main/java/me/lucko/fabric/api/permissions/v0/OfflinePermissionCheckEvent.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * This file is part of fabric-permissions-api, licensed under the MIT License.
+ *
+ *  Copyright (c) lucko (Luck) <luck@lucko.me>
+ *  Copyright (c) contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.fabric.api.permissions.v0;
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.fabricmc.fabric.api.util.TriState;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Simple permissions check event for (potentially) offline players.
+ */
+public interface OfflinePermissionCheckEvent {
+
+    Event<OfflinePermissionCheckEvent> EVENT = EventFactory.createArrayBacked(OfflinePermissionCheckEvent.class, (callbacks) -> (uuid, permission) -> {
```

Comment:
A potentially nicer way of implementing this is:

```java
Event<OfflinePermissionCheckEvent> EVENT = EventFactory.createArrayBacked(OfflinePermissionCheckEvent.class, (callbacks) -> (uuid, permission) -> {
    CompletableFuture<TriState> res = CompletableFuture.completedFuture(TriState.DEFAULT);
    for (OfflinePermissionCheckEvent callback : callbacks) {
        res = res.thenCompose(tristate -> {
            if (tristate != TriState.DEFAULT) {
                return CompletableFuture.completedFuture(tristate);
            }
            return callback.onPermissionCheck(uuid, permission);
        });
    }
    return res;
});
```

It has the benefit of:
* not calling the listener functions unless necessary (it short circuits if a result is found)
* not having a `supplyAsync` loop with `join()`s - I think ForkJoinPools are clever enough to not block up threads in this case, but it is better to avoid it alltogether if possible.

Reply:
Just thought it would be nice not to wait for each previous listener to finish, but usually, you don't have more than one (even if the extra wait shouldn't matter), so this seems totally reasonable.